src/Orleans.Runtime/Activation/IGrainContextActivator.cs

```diff
@@ -348,15 +350,58 @@ public bool TryGetConfigurator(GrainType grainType, GrainProperties properties,
                     $"Expected: public static bool {callbackMethodName}(IInvokable req)");
             }
 
-            return method.CreateDelegate<Func<IInvokable, bool>>();
+            if (method.IsStatic)
+            {
+                return new MayInterleaveStaticPredicate(method.CreateDelegate<Func<IInvokable, bool>>());
+            }
+
+            var functionType = Expression.GetFuncType(grainType, typeof(IInvokable), typeof(bool));
+            var functionDelegate = method.CreateDelegate(functionType);
+            var predicateType = typeof(MayInterleaveInstancedPredicate<>).MakeGenericType(grainType);
+
+            return Activator.CreateInstance(predicateType, functionDelegate) as IMayInterleavePredicate;
```

Comment:
(side note) at some point we will probably need to make this more AOT friendly, which will involve reworking all of this.